service/agentconf.go

```diff
@@ -323,3 +328,59 @@ func startAgent(name string, kind AgentKind, dataDir string, series string) (err
 func serviceName(agent string) string {
 	return "jujud-" + agent
 }
+
+// WriteServiceFile writes the service life in standard '/lib/systemd/' location - used by upgrade.
+func (s *systemdServiceManager) WriteServiceFile() error {
+	var lastError error
+
+	hostSeries, err := series.HostSeries()
+	if err != nil {
+		return err
+	}
+	dataDir, err := paths.DataDir(hostSeries)
+	if err != nil {
+		return err
+	}
+
+	// find the agents.
+	machineAgent, unitAgents, failedAgentNames, err := s.FindAgents(dataDir)
+	if err != nil {
+		return err
+	} else {
+		for _, name := range failedAgentNames {
+			logger.Infof("%s is not of type Machine nor Unit, ignoring", name)
+		}
+	}
+
+	for _, agentName := range append(unitAgents, machineAgent) {
+		conf, er := s.CreateAgentConf(agentName, dataDir, hostSeries)
+		if er != nil {
+			errors.Trace(err)
```

Comment:
Should this be 'err' or 'er'?

Reply:
This should be lastError = errors.Trace(err) - errors.Trace returns a new error that wraps the original one, so unless you assign it the location information it captures is lost.
(Same for the other two places below.)